### PR TITLE
New Snippet: Binary Search

### PR DESCRIPTION
**Pull request by alikuxac**
**Please describe the changes this PR makes and why it should be merged:**
I added new snippet called **Binary Search**.
**Status**
Finished, no error

- [x] Code changes have been tested on an instance of YAGPDB, or there are no code changes
- [x] I have read and followed the [contribution guide](../CONTRIBUTING.md)

### CODE REVIEW

**Comment by soudblox** [comment]
owo what can u do with this

File: snippets/binarySearch.go.tmpl [hunk 1]
```diff
@@ -0,0 +1,37 @@
+{{/*
+    This CC is a snippet help search a given element x in sorted array. The array must be ascending.
```
Comment (Jo3-L) [inline snippets/binarySearch.go.tmpl:2]: might wanna note that binary search is O(log n) which is its big selling point

File: snippets/binarySearch.go.tmpl [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            {{- .Set "Right" (sub $mid 1) -}}{{- template "binary_search" . -}}
+        {{- else -}}
+            {{- .Set "Left" (add $mid 1) -}}{{- template "binary_search" . -}}
+        {{- end -}}
+    {{else}}
+        {{- .Set "Result" -1 -}}{{- .Set "Found" false -}}
+    {{end}} 
+{{ end }}
+{{ end }}
+{{ $data = $data.Get "Result" }}
```
Comment (Jo3-L) [inline snippets/binarySearch.go.tmpl:30]:
```suggestion
{{ $data = $data.Result }}
```

File: snippets/binarySearch.go.tmpl [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    The array input is $arr. Lenght of array is $length. Element need to search is $element
+
+    Made by Alikuxac#4177
+*/}}
+
+{{ $arr := cslice 2 5 6 7 9 10 15 }}
+{{ $element := 15 }}
+{{ $length := len $arr }}
+{{ $data := sdict "Found" true "List" $arr "Left" 0 "Right" (sub $length 1) "Value" $element }}
+{{ block "binary_search" $data}}
```
Comment (Jo3-L) [inline snippets/binarySearch.go.tmpl:12]: should probably change this to `define` and `template`, clearer for examples

File: snippets/binarySearch.go.tmpl [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    This CC is a snippet help search a given element x in sorted array. The array must be ascending.
+    The array input is $arr. Lenght of array is $length. Element need to search is $element
+
+    Made by Alikuxac#4177
+*/}}
+
+{{ $arr := cslice 2 5 6 7 9 10 15 }}
+{{ $element := 15 }}
+{{ $length := len $arr }}
+{{ $data := sdict "Found" true "List" $arr "Left" 0 "Right" (sub $length 1) "Value" $element }}
```
Comment (Jo3-L) [inline snippets/binarySearch.go.tmpl:11]: probably want to default to those values for `Left` and `Right` in the template definition if they're not present so users don't have to provide it every time

**Review by Jo3-L — CHANGES_REQUESTED** [review]
lgtm other than some small nits

File: snippets/binarySearch.go.tmpl (hunk 2 above)
Comment (alikuxac) [reply snippets/binarySearch.go.tmpl]: Fixed

File: snippets/binarySearch.go.tmpl (hunk 3 above)
Comment (alikuxac) [reply snippets/binarySearch.go.tmpl]: I changed

File: snippets/binarySearch.go.tmpl (hunk 4 above)
Comment (alikuxac) [reply snippets/binarySearch.go.tmpl]: Okay, fixed

**Review by Jo3-L — APPROVED** [review]

**Pull request merged**